gpu4pyscf/df/df.py

```diff
@@ -240,47 +238,41 @@ def cholesky_eri_gpu(intopt, mol, auxmol, cd_low,
     npairs_per_ctr = np.array(npairs_per_ctr)
     total_task_list = np.argsort(npairs_per_ctr)
     task_list_per_device = []
-    for device_id in range(num_gpus):
-        task_list_per_device.append(total_task_list[device_id::num_gpus])
+    for device_id in range(_num_devices):
+        task_list_per_device.append(total_task_list[device_id::_num_devices])
 
     cd_low_f = cupy.array(cd_low, order='F', copy=False)
     cd_low_f = tag_array(cd_low_f, tag=cd_low.tag)
 
-    for gpu_id in range(num_gpus):
+    for gpu_id in range(_num_devices):
         cupy.cuda.Device(gpu_id).synchronize()
 
-    threads = []
-    for device_id in range(num_gpus):
-        task_list = task_list_per_device[device_id]
-        thread = threading.Thread(target=_cderi_task,
-                                  args=(intopt, cd_low_f, task_list, device_id, _cderi),
-                                  kwargs={"omega":omega, "sr_only":sr_only})
-        thread.start()
-        threads.append(thread)
+    futures = []
+    with ThreadPoolExecutor(max_workers=_num_devices) as executor:
+        for device_id in range(_num_devices):
+            task_list = task_list_per_device[device_id]
+            future = executor.submit(_cderi_task, intopt, cd_low_f, task_list, _cderi,
+                                     omega=omega, sr_only=sr_only, device_id=device_id)
+            futures.append(future)
```

Comment:
calls `future.result()` for each Future object. Exceptions in threads, if any,  are raised in this thread synchronization operation. This checks the. The same for the other Thread executors.